lib/Sema/CSSolver.cpp

```diff
@@ -713,24 +713,22 @@ bool ConstraintSystem::tryTypeVariableBindings(
           newBindings.push_back({subtype, binding.Kind, binding.BindingSource});
       }
 
-      if (binding.Kind == AllowedBindingKind::Subtypes) {
-        // If we were unsuccessful solving for T?, try solving for T.
-        if (auto objTy = type->getOptionalObjectType()) {
-          if (exploredTypes.insert(objTy->getCanonicalType()).second) {
-            // If T is a type variable, only attempt this if both the
-            // type variable we are trying bindings for, and the type
-            // variable we will attempt to bind, both have the same
-            // polarity with respect to being able to bind lvalues.
-            if (auto otherTypeVar = objTy->getAs<TypeVariableType>()) {
-              if (typeVar->getImpl().canBindToLValue() ==
-                  otherTypeVar->getImpl().canBindToLValue()) {
-                newBindings.push_back(
-                    {objTy, binding.Kind, binding.BindingSource});
-              }
-            } else {
+      // If we were unsuccessful solving for T?, try solving for T.
+      if (auto objTy = type->getOptionalObjectType()) {
```

Comment:
I guess the reason why it was only allowed for subtypes is related to value-to-optional promotion, which makes `T` a subtype of `T?` effectively? I think @rudkx should have more input regarding if this is for all types of bindings...

Reply:
Right, it's valid to go T? to T for subtype bindings. It's _not_ valid to go T? to T for any others, but binding the type variable to T (generally the type variable for a param) after T? has already failed gives us the opportunity in attempting to solve the system to match the T? value to the T type variable by adding a `MissingOptionalUnwrapFailure`. 

So the result is a solution with a fix, instead of unsolved.

Reply:
I think that might be reasonable to extend a check to `subtype || shouldAttemptFixes()` and leave comment about why it’s that way for posterity, WDYT?

Reply:
Yep, you're right, the change definitely should be gated off of `shouldAttemptFixes()`